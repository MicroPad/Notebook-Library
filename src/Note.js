const xml2js = require('xml2js');
const format = require('date-fns/format');
const parseDate = require('date-fns/parse');

exports.Note = function (title, time, addons) {
	this.parent = undefined;
	this.title = removeHTML(title);
	this.time = Date.parse(time);
	this.addons = addons;
	this.bibliography = [];
	this.elements = [];
	this.internalRef = generateGuid();
};

exports.Note.prototype.addSource = function (id, item, content) {
	this.bibliography.push({
		id: id,
		item: item,
		content: content
	});
};

exports.Note.prototype.addElement = function (type, args, content) {
	if (!type) return;
	if (!content || content.length === 0) return;
	this.elements.push({
		type: type,
		args: args,
		content: content
	});
};

exports.Note.prototype.search = function (query) {
	var pattern = new RegExp("\\b" + query.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&"), 'i');
	if (pattern.test(this.title)) return this;

	if (query.length > 1 && query.charAt(0) == '#') {
		pattern = new RegExp("(^|\\s)" + query.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&") + "(\\b)", 'i');
		for (var i = 0; i < this.elements.length; i++) {
			var el = this.elements[i];
			if (el.type !== "markdown") continue;
			if (pattern.test(el.content)) return this;
		}
	}
	return false;
};

exports.Note.prototype.toXMLObject = function () {
	var parseableNote = {
		note: {
			$: {
				title: removeHTML(this.title),
				time: format(parseDate(this.time, 'YYYY-MM-DDTHH:mm:ss.SSSZ'))
			},
			addons: [],
			bibliography: []
		}
	};

	var imports = {
		import: []
	};
	for (let k in this.addons) {
		imports.import.push(this.addons[k]);
	}
	parseableNote.note.addons.push(imports);

	var sources = {
		source: []
	};
	for (let k in this.bibliography) {
		var source = this.bibliography[k];
		sources.source.push({
			_: source.content,
			$: {
				id: source.id,
				item: source.item
			}
		});
	}
	parseableNote.note.bibliography.push(sources);

	let elements = {};
	for (let k in this.elements) {
		var element = this.elements[k];
		if (!elements[element.type]) elements[element.type] = [];

		var elementToPush = {
			_: element.content,
			$: {}
		};

		for (let argName in element.args) {
			elementToPush.$[argName] = element.args[argName];
		}

		elements[element.type].push(elementToPush);
	}
	for (let k in elements) {
		var elementGroup = elements[k];
		parseableNote.note[k] = elementGroup;
	}

	return parseableNote;
};

exports.Note.prototype.toXML = function () {
	var builder = new xml2js.Builder({
		allowSurrogateChars: true,
		headless: true,
		renderOpts: {
			'pretty': false
		},
		cdata: true
	});
	return builder.buildObject(this.toXMLObject());
};

exports.Note.prototype.toMarkdown = function (assets) {
	var mdNote = "";
	for (var i = 0; i < this.elements.length; i++) {
		var element = this.elements[i];
		var citation = "";
		for (var j = 0; j < this.bibliography.length; j++) {
			var source = this.bibliography[j];

			if (source.item === element.args.id) {
				citation = "[[{0}]]({1})".format(source.id, source.content);
			}
		}

		switch (element.type) {
			case "markdown":
				mdNote += element.content + citation + "\n\n";
				break;

			case "drawing":
			case "image":
				var content = "";
				if (element.args.ext) {
					content = assets[element.args.ext];
				} else {
					content = element.content;
				}

				mdNote += "![]({0}){1}\n\n".format(content, citation);
				break;

			case "file":
			case "recording":
				var content = "";
				if (element.args.ext) {
					content = assets[element.args.ext];
				} else {
					content = element.content;
				}

				mdNote += "[{0}]({1}){2}\n\n".format(element.args.filename, content, citation);
				break;
		}
	}

	return { title: this.title, md: mdNote };
};

exports.Note.prototype.getUsedAssets = function () {
	var usedAssets = [];
	for (var i = 0; i < this.elements.length; i++) {
		var element = this.elements[i];
		if (element.content === "AS") usedAssets.push(element.args.ext);

		if (element.type === "markdown") {
			var inlineImages = [];
			element.content.replace(/!!\(([^]+?)\)/gi, (match, p1) => {
				inlineImages.push(p1);
			});
			element.content.replace(/!!\[([^]+?)\]/gi, (match, p1) => {
				inlineImages.push(p1);
			});

			if (inlineImages) {
				let uuid = inlineImages[i];
				if (uuid) usedAssets.push(uuid);
			}
		}
	}

	return usedAssets;
};

// Thanks to http://stackoverflow.com/a/4673436/998467
if (!String.prototype.format) {
	String.prototype.format = function () {
		var args = arguments;
		return this.replace(/{(\d+)}/g, function (match, number) {
			return typeof args[number] != 'undefined' ? args[number] : match;
		});
	};
}

//Thanks to https://stackoverflow.com/a/105074
function generateGuid() {
	function s4() {
		return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
	}
	return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4();
}

function removeHTML(string) {
	return string.replace(/<[^>]*>/, "");
}