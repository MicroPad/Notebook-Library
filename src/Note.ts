import { NPXObject } from './NPXObject';
import { format } from 'date-fns';
import { Asset } from './index';

export type NoteElement = {
	type: 'markdown' | 'image' | 'drawing' | 'file' | 'recording';
	content: string;
	args: ElementArgs;
};

export type ElementArgs = {
	id: string;
	x: string;
	y: string;
	width?: string;
	height?: string;
	fontSize?: string;
	filename?: string;
	ext?: string;
};

export type Source = {
	id: number;
	item: string;
	content: string;
};

export type MarkdownNote = {
	title: string;
	md: string;
};

export default class Note extends NPXObject {
	constructor(
		public readonly title: string,
		public readonly time: number = new Date().getTime(),
		public readonly elements: NoteElement[] = [],
		public readonly bibliography: Source[] = [],
		internalRef?: string
	) {
		super(title, internalRef);
	}

	public addElement(element: NoteElement): Note {
		return this.clone({
			elements: [
				...this.elements,
				element
			]
		});
	}

	public addSource(source: Source): Note {
		return this.clone({
			bibliography: [
				...this.bibliography,
				source
			]
		});
	}

	public search(query: string): Note[] {
		// Title search
		let pattern = new RegExp("\\b" + query.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, '\\$&'), 'i');
		if (pattern.test(this.title)) return [this];

		// Hashtag search
		if (query.length > 1 && query.charAt(0) === '#') {
			pattern = new RegExp("(^|\\s)" + query.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&") + "(\\b)", 'i');

			// Check if any markdown elements contain the hashtag
			if (
				this.elements
					.filter(e => e.type === 'markdown')
					.some(e => pattern.test(e.content))
			) return [this];
		}

		return [];
	}

	public toXmlObject(): any {
		const elements = {};
		this.elements.forEach(e => {
			if (!elements[e.type]) elements[e.type] = [];
			elements[e.type].push({
				$: { ...e.args },
				_: e.content
			});
		});

		const bibliography = this.bibliography.map(source => {
			return {
				source: {
					$: {
						id: source.id,
						item: source.item
					},
					_: source.content
				}
			};
		});

		return {
			note: {
				$: {
					title: this.title,
					time: format(this.time, 'YYYY-MM-DDTHH:mm:ss.SSSZ')
				},
				addons: [[]], // We aren't supporting addons in v3 of the parser but we'll keep this for NPX compatibility
				bibliography: (bibliography.length > 0) ? bibliography : [[]],
				...elements
			}
		};
	}

	public async toMarkdown(assets: Asset[]): Promise<MarkdownNote> {
		const assetMap = {};
		assets.forEach(a => assetMap[a.uuid] = a);

		const md: string = (await Promise.all(this.elements
			.filter(e => ['markdown', 'drawing', 'image'].includes(e.type))
			.map(async e => {
				let md: string | undefined;

				switch (e.type) {
					case 'markdown':
						md = e.content + '\n\n';
						break;

					case 'drawing':
					case 'image':
						const asset = assetMap[e.args.ext || 0];
						if (!asset) return '';

						md = `![](${await asset.toString()})\n\n`;
						break;
				}
				if (!md) return '';

				// Bibliography
				const bib = this.bibliography
					.filter(s => s.item === e.args.id)
					.map(s => s.content);

				if (bib.length > 0) {
					md += `***Bibliography***  \n`;
					md += bib
						.map(content => `- <${content}>\n`)
						.reduce((str, source) => str += source, '') + '\n';
				}

				return md;
			})))
			.reduce((str, elementMd) => str += elementMd, '');

		return { title: this.title, md };
	}

	public clone(opts: Partial<Note> = {}): Note {
		return new Note(
			opts.title || this.title,
			opts.time || this.time,
			opts.elements || this.elements,
			opts.bibliography || this.bibliography,
			opts.internalRef || this.internalRef
		);
	}
}
