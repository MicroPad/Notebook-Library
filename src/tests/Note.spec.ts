import { TestUtils } from './TestUtils';
import { Note } from '../index';
import { ElementArgs, NoteElement, Source } from '../Note';

describe('Note', () => {
	let element: NoteElement;
	let source: Source;

	beforeEach(() => {
		element = {
			content: 'Hello, World!',
			type: 'markdown',
			args: {
				id: 'markdown1',
				x: '10px',
				y: '10px'
			}
		};

		source = {
			id: 1,
			item: 'markdown1',
			content: 'https://nick.geek.nz'
		};
	});

	it('should construct', () => {
		// Arrange
		const title = 'test';

		// Act
		const res = TestUtils.makeNote(title);

		// Assert
		expect(res).toBeInstanceOf(Note);
		expect(res.title).toEqual(title);
	});

	describe('addElement', () => {
		let note: Note;

		beforeEach(() => {
			note = TestUtils.makeNote('test note');
		});

		it('should add a new element', () => {
			//Arrange
			// Act
			const res = note.addElement(element);

			// Assert
			expect(res.elements[0]).toEqual(element);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = note.addElement(element);

			// Assert
			expect(res).not.toBe(note);
		});
	});

	describe('addSource', () => {
		let note: Note;

		beforeEach(() => {
			note = TestUtils.makeNote('test note');
		});

		it('should add a new source', () => {
			//Arrange
			// Act
			const res = note.addSource(source);

			// Assert
			expect(res.bibliography[0]).toEqual(source);
		});

		it('should create a new object', () => {
			//Arrange
			// Act
			const res = note.addSource(source);

			// Assert
			expect(res).not.toBe(note);
		});
	});

	describe('search', () => {
		describe('title search', () => {
			it('should return empty array on no match', () => {
				// Arrange
				const note = TestUtils.makeNote('test');

				// Act
				const res = note.search('invalid');

				// Assert
				expect(res).toEqual([]);
			});

			it('should return an array with the note on a match', () => {
				// Arrange
				const note = TestUtils.makeNote('test');

				// Act
				const res = note.search('te');

				// Assert
				expect(res).toEqual([note]);
			});

			it('should return an array with the note on an empty query', () => {
				// Arrange
				const note = TestUtils.makeNote('test');

				// Act
				const res = note.search('');

				// Assert
				expect(res).toEqual([note]);
			});
		});

		describe('hashtag search', () => {
			it('should return empty array on no match', () => {
				// Arrange
				const note = TestUtils.makeNote('test');

				// Act
				const res = note.search('#test');

				// Assert
				expect(res).toEqual([]);
			});

			it('should not accept a partial match', () => {
				// Arrange
				let note = TestUtils.makeNote('test');
				note = note.addElement({
					type: 'markdown',
					args: {} as ElementArgs,
					content: '#match'
				});

				// Act
				const res = note.search('#mat');

				// Assert
				expect(res).toEqual([]);
			});

			it('should return an array with the note on a full match', () => {
				// Arrange
				let note = TestUtils.makeNote('test');
				note = note.addElement({
					type: 'markdown',
					args: {} as ElementArgs,
					content: '#match'
				});

				// Act
				const res = note.search('#match');

				// Assert
				expect(res).toEqual([note]);
			});
		});
	});

	it('should generate XML Object with required data', () => {
		// Arrange
		let note = TestUtils.makeNote('test note');
		note = note.addElement(element);
		note = note.addSource(source);

		// Act
		const res = note.toXmlObject();

		// Assert
		expect(res).toMatchSnapshot();
	});
});
