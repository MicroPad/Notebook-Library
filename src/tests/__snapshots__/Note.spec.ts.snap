// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note should generate XML Object with required data 1`] = `
Object {
  "note": Object {
    "$": Object {
      "time": "1970-01-01T12:00:00.001+12:00",
      "title": "test note",
    },
    "addons": Array [
      Array [],
    ],
    "bibliography": Array [
      Object {
        "source": Object {
          "$": Object {
            "id": 1,
            "item": "markdown1",
          },
          "_": "https://nick.geek.nz",
        },
      },
    ],
    "markdown": Array [
      Object {
        "$": Object {
          "id": "markdown1",
          "x": "10px",
          "y": "10px",
        },
        "_": "Hello, World!",
      },
    ],
  },
}
`;
